exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/SqlPatternContainsMatcher.java

```diff
@@ -19,44 +19,283 @@
 
-      return 1;
+    private Matcher3() {
+      super();
     }
 
-    return  0;
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
+      final int lengthToProcess  = end - start -2;
+      final byte firstPattByte   = patternArray[0];
+      final byte secondPattByte  = patternArray[1];
+      final byte thirdPattByte   = patternArray[2];
+
+      // simplePattern string has meta characters i.e % and _ and escape characters removed.
+      // so, we can just directly compare.
+      for (int idx = 0; idx < lengthToProcess; idx++) {
+        final byte inputByte = drillBuf.getByte(start + idx);
+
+        if (firstPattByte != inputByte) {
+          continue;
+        } else {
+          final byte secondInByte = drillBuf.getByte(start + idx +1);
+          final byte thirdInByte  = drillBuf.getByte(start + idx +2);
+
+          if (secondInByte == secondPattByte && thirdInByte == thirdPattByte) {
+            return 1;
+          }
+        }
+      }
+      return 0;
+    }
+  }
+
+  /** Handles patterns with arbitrary length */
+  private final class MatcherN extends MatcherFcn {
+
+    private MatcherN() {
+      super();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
+
+      if (patternLength == 0) {
+        return 1;
+      }
+
+      final int lengthToProcess = end - start - patternLength + 1;
+      int patternIndex          = 0;
+      final byte firstPattByte  = patternArray[0];
+
+      // simplePattern string has meta characters i.e % and _ and escape characters removed.
+      // so, we can just directly compare.
+      for (int idx = 0; idx < lengthToProcess; idx++) {
+        final byte inputByte = drillBuf.getByte(start + idx);
+
+        if (firstPattByte == inputByte) {
+          for (patternIndex = 1; patternIndex < patternLength; ++patternIndex) {
+            final byte currInByte   = drillBuf.getByte(start + idx + patternIndex);
+            final byte currPattByte = patternArray[patternIndex];
+
+            if (currInByte != currPattByte) {
+              break;
+            }
+          }
+
+          if (patternIndex == patternLength) {
+            break;
+          }
+        }
+      }
+      return patternIndex == patternLength ? 1 : 0;
+    }
+  }
+
+  /**
+   * Boyer-Moore matcher algorithm; excellent for large patterns and for prefix patterns which appear
+   * frequently in the input.
+   */
+  private final class BoyerMooreMatcher extends MatcherFcn {
+    private final int[] offsetTable;
+    private final int[] characterTable;
+
+    private BoyerMooreMatcher() {
+      super();
+
+      this.offsetTable    = makeOffsetTable();
+      this.characterTable = makeCharTable();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected int match(int start, int end, DrillBuf drillBuf)  {
+      if (patternLength == 0) {
```

Comment:
do we even come here if patternLength is 0 ?

Reply:
Yes, it would have come here; but for simplicity I guess I can handle that use-case within the switch case.